### PR TITLE
Substitute a replacement character for invalid UTF-16 text in a TextSpan

### LINKED ISSUE #84693: REPLACEMENT CHARACTER (�) for not well-formed UTF-16 character
## Use case

Currently `Text` widget needs well-formed UTF-16 text to render, and throws `string is not well-formed UTF-16` error if its not

```dart
Text('😈'.substring(0,1))
```

it's very dangerous, consider the following scenario

- we have an application and let users to leave comments
- a hacker leave a comment with not well-formed UTF-16 character
- other users cannot open app because it crashes with `string is not well-formed UTF-16` error

## Proposal

Simply use [Specials (Unicode block)](https://en.wikipedia.org/wiki/Specials_(Unicode_block))

> U+FFFD � REPLACEMENT CHARACTER used to replace an unknown, unrecognized, or unrepresentable character

� is the most common used character to ignore not well-formed UTF-16 characters

### PR DESCRIPTION
**Pull request by jason-simmons**
Fixes https://github.com/flutter/flutter/issues/84693

### CODE REVIEW

File: packages/flutter/test/painting/text_painter_test.dart [hunk 1]
```diff
@@ -952,6 +952,22 @@ void main() {
...
+      exception = details.exception;
+    };
+
+    final TextPainter painter = TextPainter()
+      ..textDirection = TextDirection.ltr;
+
+    String text = 'Hello\uD83DWorld';
+    painter.text = TextSpan(text: text);
+    painter.layout();
+    expect(painter.width, greaterThan(0.0));
```
Comment (Hixie) [inline packages/flutter/test/painting/text_painter_test.dart:968]: it should be exactly 10.0, right? assuming default font size is 10 and font is Ahem

**Review by Hixie — APPROVED** [review]
LGTM, but travis is angry.

...I really should update this image.

**Comment by Hixie** [comment]
Bonus points if you mention this somewhere in the API docs, maybe for Text, TextSpan, and TextPainter in particular.

File: packages/flutter/test/painting/text_painter_test.dart (hunk 1 above)
Comment (jason-simmons) [reply packages/flutter/test/painting/text_painter_test.dart]: done

**Comment by sm2017** [comment]
@jason-simmons in [this line](https://github.com/flutter/flutter/pull/84887/commits/6fbae3f1655b71e1596ebb6d98820050850c9891#diff-f448a03ba39ca8c93d63bd76dc0421c08787b2607caa58bfd8ea9a28d0a30b2aR258), As I understand you replace whole text with REPLACEMENT CHARACTER, Why whole text? We should replace only invalid characters

**Comment by Hixie** [comment]
Replacing individual characters is a lot of work, I really don't think we should do that given that the whole point is to fail. If you want to replace invalid characters, that's something to do in your app.

**Comment by sm2017** [comment]
@Hixie I understand,  Assume I want to do it, There is many many string and `Text` in the application, Some `Text` are in the dependencies of dependencies
It's very cumbersome to do it in application layer, At least the must be a `Global` option to override text sanitization

**Comment by Hixie** [comment]
You would have to provide a central place in your application where strings were sanitized, yes.

But fundamentally, strings should be sanitized long before they reach `Text`. If they're not, that indicates a more fundamental problem. I don't think it's Flutter's job to provide APIs to make it easier to work around such fundamental problems.

**Pull request merged**